### PR TITLE
support for multi-threaded providers on tokio runtime

### PR DESCRIPTION
**Pull request by stevelr**
Support providers using tokio runtime and async.
logging module logs through crossbeam channel to a logger on dedicated logging thread.
added hooks to make it possible to mock host with shell script sending rpc and control messages w/ json.

### CODE REVIEW

**Comment by autodidaddict** [comment]
I have to admit that I don't understand the need for the logging module. All of the regular logging macros work from inside background threads so I'm trying to figure out why we need the complexity of our own custom logger?

File: rpc-rs/src/channel_log.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+// an internally-used value to help vertically align consecutive log messsages.
+static MAX_MODULE_WIDTH: AtomicUsize = AtomicUsize::new(0);
+
+// All log lines have a newline appended.
+// when running the provider stand-alone on linux, '\n' is sufficient,
+// but when running under elixir, on linux, '\r\n' was needed for the
+// error formatting to look right. Haven't tried other platforms yet,
+// but the difference between linux-standalone and linux-in-elixir
+// means querying the os for line endings won't always give the correct answer.
```
Comment (brooksmtownsend) [inline rpc-rs/src/channel_log.rs:198]: `\r\n` is needed for Windows too 😄

**Review by brooksmtownsend — COMMENTED** [review]
There's a good bit to grok here so I have a few questions:

1. It looks like all provider functions are async now, will provider code always be in an asynchronous context? Wouldn't want to cause pain at the sync/async barrier in provider code
2. Is the receiver of the logs on the host side, or on the provider side? I think I'm curious about what @autodidaddict said too, I think this looks great and would be resilient for bursty log writes, and I wonder if this is more complicated than needed.
3. Love the codegen abstractions here (load_host_data, HostBridge::new()). Am I correct in thinking that the flow for a developer would be something like declaring that the provider implements the provider trait, and it gets the `HostData` and nats connections for free (no user declaration of a NATS client)?

**Comment by stevelr** [comment]
I'm not crazy about it either. I didn't have reliable logging until adding the channel and sending everything to a single thread, even when running `try_init` inside every new thread at the time it was created. I don't know if the problem was a lock on stderr, or thread local data, or something else. I plan to do more testing later, and if it's not necessary, I'll remove it. 

There was one other minor benefit to having the logger. For some reason, when the provider is loaded by the wasmcloud host process, an extra `\r` was required for the logs to print correctly. If you logged "hello" five times it would look like this:
```
hello
     hello
          hello
               hello
                    hello
```
 adding `\r` to the quoted strings inside debug/info/error macros didn't seem to help. This formatting issue is unique to being a child process of wasmcloud host, because if I run the provider straight from the linux command line, the log lines are formatted correctly.
If we didn't have the logger I'm sure we could figure out the \r issue, so I'm not saying this motivates a whole logger, but since the logger was there, it provided an easy fix.

File: rpc-rs/src/channel_log.rs (hunk 1 above)
Comment (stevelr) [reply rpc-rs/src/channel_log.rs:198]: right, but println!() is supposed to do the right thing for your platform so if you use that, then you don't have to worry about whether you are on windows or mac or linux. The issue _seems to be_  that when elixir pipes the stdout/stderr of the process, it goes to a console that needs '\r' added, even on linux. A provider _should_ check the OS to make line endings correct. a provider _should not_ have to check whether it's running as a subprocess of elixir before determining its line endings.

**Comment by stevelr** [comment]
> It looks like all provider functions are async now, will provider code always be in an asynchronous context? Wouldn't want to cause pain at the sync/async barrier in provider code

All modern http servers are async. If it were sync, it would need one thread per http request, and the thread would block to send the request over nats to the actor and wait for the response. That would be ok for a low-activity http server, if you're running it at home for a home automation project, but it wouldn't scale well at all. 
I would predict that the default is that other providers will want to be async. If a provider wants to use synchronous methods (other than blocking io) inside an async function, it's easy - just call the synchronous method. It's harder the other way around (if the handler was not async and they wanted to call an async function), so async as default might actually be easier. If the developer wants to do blocking io calls, though, they should wrap it in another tokio function. I'll open an issue to make sure we develop example code that demonstrates how to do blocking io calls inside a provider.

**Comment by stevelr** [comment]
> Is the receiver of the logs on the host side, or on the provider side?

It's in a thread in the provider OS process, launched when the provider starts.

**Comment by stevelr** [comment]
> Love the codegen abstractions here (load_host_data, HostBridge::new()). Am I correct in thinking that the flow for a developer would be something like declaring that the provider implements the provider trait, and it gets the HostData and nats connections for free (no user declaration of a NATS client)?

There's one function that returns a nats client builder, returns it to the provider, and the provider can modify this and then passes it to the HostData connect. So it takes one line for the provider if they don't want to modify the nats configuration or use nats, but I exposed it at that point so that the provider can
- change the nats connect parameters, for example, if they want to wrap it in tls
- use the nats connection for anything else in their app. This might be a minor point because they could also open their own nats connection, but if they re-use the same nats client, then shutdown is cleaner.

**Pull request merged**